spec/unit/plugins/linux/block_device_spec.rb

```diff
@@ -0,0 +1,74 @@
+#
+#  Author:: Jennifer Marie Howard-Brown <jm.howardbrown@gmail.com>
+#  License:: Apache License, Version 2.0
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+require File.expand_path(File.dirname(__FILE__) + '/../../../spec_helper.rb')
+
+describe Ohai::System, 'Linux Block Device Plugin' do
+  DISKS = {
+    'sda' => {
+        "size" => "7814037168",
+        "removable" => "0",
+        "model" => "WDC WD4000F9YZ-0",
+        "rev" => "1A01",
+        "state" => "running",
+        "timeout" => "30",
+        "queue_depth" => "1",
+        "vendor" => "ATA"
+    },
+    'dm-0' => {
+        "size" => "7806976",
+        "removable" => "0"
+    }
+  }
+
+  before(:each) do
+    @plugin = get_plugin('linux/block_device')
+    allow(@plugin).to receive(:collect_os).and_return(:linux)
+    
+
+    allow(File).to receive(:exists?).with("/sys/block").and_return(true)
+    allow(Dir).to receive(:[]).with ("/sys/block/*") do
+      DISKS.collect { |disk, files| "/sys/block/#{disk}" }
+    end
+
+    DISKS.each do |disk, checks|
+      allow(File).to receive(:open).with(Regexp.new(disk)) do |arg|
+        filepath = arg.split('/')
+        tmp_double = double(arg)
+        allow(tmp_double).to receive(:read_nonblock).and_return(checks[filepath.last])
+        allow(tmp_double).to receive(:close)
+        tmp_double
+      end
```

Comment:
I think you're making your life harder here by passing the block in to the allow.  You should really just create a double and then ideally revert your changes to the code and the yield the double.  In order to reduce the code duplication it would probably be better to create a method to handle parameterizing the return value of `:read_nonblock`.